libs/pilight/hardware/raspyrfm.c

```diff
@@ -0,0 +1,441 @@
+		logprintf(LOG_ERR, "no gpio-platform configured");
+		return EXIT_FAILURE;
+	}
+	if(wiringXSetup(platform, _logprintf) < 0) {
+		FREE(platform);
+		return EXIT_FAILURE;
+	}
+	FREE(platform);
+
+	if (rfmsettings.spi_ch == -1)
+		return EXIT_FAILURE;
+
+	wiringXSPISetup(rfmsettings.spi_ch, 2500000UL);
+	
+	//send config to raspyrfm
+	for(i=0; i<sizeof(rfmcfg) / sizeof(rfmcfg[0]); i++)
+		writeReg(rfmcfg[i].reg, rfmcfg[i].val);
+		
+	//set frequency
+	if (rfmsettings.frequency == 0)
+		return EXIT_FAILURE;
+
+	freqword = FREQTOFREG(rfmsettings.frequency); //MHz
+ 
+	writeReg(REGFR, (freqword >> 16) & 0xFF);
+	writeReg(REGFR + 1, (freqword >> 8) & 0xFF);
+	writeReg(REGFR + 2, freqword & 0xFF);
+	
+	startReceive();
+
+#ifdef PILIGHT_UNITTEST
+	gpio_dio0 = 1;
+#else
+	if(rfmsettings.spi_ch == 0)
+		gpio_dio0 = 6;
+	else
+		gpio_dio0 = 5;
+#endif
+	
+	if(wiringXISR(gpio_dio0, ISR_MODE_RISING) < 0) {
+		logprintf(LOG_ERR, "unable to register interrupt for pin %d", 6);
+		return EXIT_FAILURE;
+	}
+	if((poll_req = MALLOC(sizeof(uv_poll_t))) == NULL) {
+		OUT_OF_MEMORY /*LCOV_EXCL_LINE*/
+	}
+	int fd = wiringXSelectableFd(gpio_dio0);
+
+	uv_poll_init(uv_default_loop(), poll_req, fd);
+
+#ifdef PILIGHT_UNITTEST
+	char *dev = getenv("PILIGHT_RASPYRFM_READ");
+	if(dev == NULL) {
+		pollpri = UV_PRIORITIZED; /*LCOV_EXCL_LINE*/
+	} else {
+		pollpri = UV_READABLE;
+	}
+#endif
+	uv_poll_start(poll_req, pollpri, poll_cb);
+	
+	eventpool_callback(REASON_SEND_CODE, raspyrfmSend, 0);
+	return EXIT_SUCCESS;
+#else
+	logprintf(LOG_ERR, "the raspyrfm module is not supported on this hardware");
+	return EXIT_FAILURE;
+#endif
+}
+
+static unsigned short setFrequency(int freqkHz) {
+	rfmsettings.frequency = freqkHz / 1000.0;
+	if((rfmsettings.frequency > 800) && (rfmsettings.frequency < 900))
+		raspyrfm->hwtype=RF868;
+	else if ((rfmsettings.frequency > 430) && (rfmsettings.frequency < 440))
+		raspyrfm->hwtype=RF433;
+	else {
+		raspyrfm->hwtype=NONE;
+		return EXIT_FAILURE;
+	}
+
+	return EXIT_SUCCESS;
+}
+
+static unsigned short raspyrfmSettings(struct JsonNode *json) {
+	if(strcmp(json->key, "frequency") == 0) {
+		if(json->tag == JSON_NUMBER) {
+			if (setFrequency((int) json->number_) != EXIT_SUCCESS)
+				return EXIT_FAILURE;
+		}
+		else
+			return EXIT_FAILURE;
+	}
+	
+	if(strcmp(json->key, "spi-channel") == 0) {
+		if(json->tag == JSON_NUMBER) {
+			rfmsettings.spi_ch = (int) json->number_;
+		} else {
+			return EXIT_FAILURE;
+		}
+	}
+	return EXIT_SUCCESS;
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void raspyrfmInit(void) {
+	hardware_register(&raspyrfm);
+	hardware_set_id(raspyrfm, "raspyrfm");
+	
+	options_add(&raspyrfm->options, "f", "frequency", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_NUMBER, NULL, "^[0-9.]+$");
+	options_add(&raspyrfm->options, "c", "spi-channel", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_NUMBER, NULL, "^[01]$");
+
+	int frequency = 0;
+	if(config_setting_get_number("frequency", 0, &frequency) != 0) {
```

Comment:
Don't forget to handle the error or actually enforce the options with the option mask.

Reply:
I missed some lines when pushing. Now frequency is checked.